### ISSUE TITLE
Ability to "block" other jobs while a specific job is running

### ISSUE THREAD

**Issue by cole-h** [zrepl/zrepl#447]
Maybe somewhat related to https://github.com/zrepl/zrepl/issues/333.

---

I have a backup pool that I upload all backups to once a day. I do this by running a systemd timer that runs `zrepl signal wakeup <job>`, which will push the backups to my backup pool. This happens at exactly midnight, but I've noticed that sometimes my actual snapshot / prune jobs will cause `PLANNING-ERROR`s and `STEP-ERROR`s as the snapshots are pruned out from underneath it (requiring me to run it again).

Basically, I'd like a way to specify "dependencies" between jobs, where if job B depends on job A, and job A is currently running, then job B will wait until job A is done before it starts snapshotting / replicating / pruning.

**Comment by InsanePrawn** [comment]
Hi there, first of, I fully agree on having a `pause` subcommand, but what you're experiencing shouldn't be a fatal error in the first place (see #349), so my question is, do we need job-to-job dependencies?

I've suggested doing some sort of in-zrepl-memory locking of datasets to avoid things exactly like your pruner problem, plus being able to specify the intervals/trigger mechanisms for replication and snapshotting separately (#397) would IMO remove the need for a lot of snapjobs anyway.
IMO a generic solution like the locking lets arbitrary jobs interoperate with (and admittedly deadlock datasets for) each other without having to be 'introduced' to one another in the config.
Would a combination of (locking and/or retries), #397, and a 'stupid' pause command for jobs solve your use case?

In case you still think a dependency system is necessary, here's some considerations:
- how do we expose making different job stages, i.e. snapping, pruning or replication depend (i.e. block?) on other jobs[' stages]? Are there that many sensible combinations?
- how do you deal with deadlocks, i.e. hung jobs. This might be an interesting question to answer in a broader sense anyway. Job timeouts?


There's also the suggestion of a `wait` subcommand (#436) that you could integrate into your systemd units, (which sound similar to my PR (#420) btw, feel free to leave feedback and/or your unit files for comparison) and track the dependency within systemd I suppose?

I've accepted the occasional sync error as part of my "snapshot + prune locally rapidly and replicate often-ish" strategy (eventual consistency, amiright?), but I see the use case for low frequency replication.

**Comment by cole-h** [comment]
It's not actually a fatal error, I just want to guarantee (almost) all of my snapshots are replicated (and having my older snapshots pruned halts the replication process fairly early on for those datasets).

Honestly, the ability to specify an interval would be fantastic. If I could guarantee that snapshots happen at exactly `00:00:00`, `00:05:00`, `00:10:00`, etc, then it would be easier to prune within those windows (especially coupled with a locking mechanism to prevent currently-replicating snapshots from getting pruned).

This is one of my gripes coming from znapzend -- all of those snapshots would happen exactly on the 5 minute mark, whereas zrepl just picks up where the previous snapshot left off, plus 5 minutes.

So yeah, locking (maybe accompanied by retries), in addition to cron-like scheduling and a `zrepl signal pause JOB` would definitely solve this problem for me.

**Comment by problame** [comment]
> Honestly, the ability to specify an interval would be fantastic. If I could guarantee that snapshots happen at exactly 00:00:00, 00:05:00, 00:10:00, etc, then it would be easier to prune within those windows (especially coupled with a locking mechanism to prevent currently-replicating snapshots from getting pruned).
> 
> This is one of my gripes coming from znapzend -- all of those snapshots would happen exactly on the 5 minute mark, whereas zrepl just picks up where the previous snapshot left off, plus 5 minutes.

I think that's a separate feature request?
```
snapshotting:
  type: periodic
  prefix: zrepl_
  # either `cron` or `interval`
  cron: "*/5 * * * *"
  ...
```

Also note that we added `zrepl signal snapshot JOB` in `2c8c2cfa14576e5e2ceac8a52be7d1f5dd41ec04` (available in zrepl 0.4-rc1) which can be used to trigger out-of-the-line snapshots like so:

```
# zrepl.yml
name: jobname
snapshotting:
  type: periodic
  prefix: zrepl
  interval: 100y
```

```
# crontab / whatever
zrepl signal snapshot "jobname"
```

**Comment by problame** [comment]
> I have a backup pool that I upload all backups to once a day. I do this by running a systemd timer that runs zrepl signal wakeup <job>, which will push the backups to my backup pool. This happens at exactly midnight, but I've noticed that sometimes my actual snapshot / prune jobs will cause PLANNING-ERRORs and STEP-ERRORs as the snapshots are pruned out from underneath it (requiring me to run it again).
> 
> Basically, I'd like a way to specify "dependencies" between jobs, where if job B depends on job A, and job A is currently running, then job B will wait until job A is done before it starts snapshotting / replicating / pruning.

Just to get this straight: you are using separate jobs for snapshotting and pruning, right?

The mid-term plan is to fully decouple snapshot management and replication which will force us to add some form of coordination between jobs, probably filesystem subtree / filesystem / snapshot-scoped locking.
However there are lots of headaches involved in distributed locking, e.g., "if replication is interrupted due to failing network link, how long should one side of the setup keep the resource locks created by the other side". I think it's going to result in some kind of lease-based system. But we'd also need some way to expose the current locking status to users, e.g. via zrepl status, so that users can understand why a job is not making progress, etc.

**Comment by cole-h** [comment]
> > Honestly, the ability to specify an interval would be fantastic. If I could guarantee that snapshots happen at exactly 00:00:00, 00:05:00, 00:10:00, etc, then it would be easier to prune within those windows (especially coupled with a locking mechanism to prevent currently-replicating snapshots from getting pruned).
> > This is one of my gripes coming from znapzend -- all of those snapshots would happen exactly on the 5 minute mark, whereas zrepl just picks up where the previous snapshot left off, plus 5 minutes.
> 
> I think that's a separate feature request?

I guess so, whoops. I meant  to be agreeing with https://github.com/zrepl/zrepl/issues/397
 
> ```
> snapshotting:
>   type: periodic
>   prefix: zrepl_
>   # either `cron` or `interval`
>   cron: "*/5 * * * *"
>   ...
> ```
> 
> Also note that we added `zrepl signal snapshot JOB` in `2c8c2cfa14576e5e2ceac8a52be7d1f5dd41ec04` (available in zrepl 0.4-rc1) which can be used to trigger out-of-the-line snapshots like so:
> 
> ```
> # zrepl.yml
> name: jobname
> snapshotting:
>   type: periodic
>   prefix: zrepl
>   interval: 100y
> ```
> 
> ```
> # crontab / whatever
> zrepl signal snapshot "jobname"
> ```

**Comment by cole-h** [comment]
Preface: I posted before I was ready somehow... So I deleted it. Here it is again, but finished :D

---

> I think that's a separate feature request?

I guess so, whoops; I meant to be agreeing with https://github.com/zrepl/zrepl/issues/397.

> ```
> snapshotting:
>   type: periodic
>   prefix: zrepl_
>   # either `cron` or `interval`
>   cron: "*/5 * * * *"
>   ...
> ```
> 
> Also note that we added `zrepl signal snapshot JOB` in `2c8c2cfa14576e5e2ceac8a52be7d1f5dd41ec04` (available in zrepl 0.4-rc1) which can be used to trigger out-of-the-line snapshots like so:
> 
> ```
> # zrepl.yml
> name: jobname
> snapshotting:
>   type: periodic
>   prefix: zrepl
>   interval: 100y
> ```
> 
> ```
> # crontab / whatever
> zrepl signal snapshot "jobname"
> ```

That definitely sounds interesting; I'll have to experiment with it some time. Does the trigger "restart" the interval? Or if I conceivably keep this system for 100+ years, will it snapshot on its own?

---

> Just to get this straight: you are using separate jobs for snapshotting and pruning, right?

Kinda, but not really? I snapshot and prune my system pool with one job, and then I just use a `push` job to push to my backup pool (which prunes the backup pool's snapshots).

You can see my config here: https://github.com/cole-h/nixos-config/blob/c7a18b25a167877c33f651e4cb3b40dd3cf02f8f/hosts/scadrial/modules/services.nix#L6-L164

> The mid-term plan is to fully decouple snapshot management and replication which will force us to add some form of coordination between jobs, probably filesystem subtree / filesystem / snapshot-scoped locking.
> However there are lots of headaches involved in distributed locking, e.g., "if replication is interrupted due to failing network link, how long should one side of the setup keep the resource locks created by the other side". I think it's going to result in some kind of lease-based system. But we'd also need some way to expose the current locking status to users, e.g. via zrepl status, so that users can understand why a job is not making progress, etc.

Yeah, that does sound tough... If it's achievable, it would greatly improve usability, IMHO. Maybe it could be opt-in by default, so users already know that some jobs may "deadlock" due to it?

**Comment by problame** [comment]
> That definitely sounds interesting; I'll have to experiment with it some time. Does the trigger "restart" the interval? Or if I conceivably keep this system for 100+ years, will it snapshot on its own?

The implementation (cc @calistoc ) starts a new timer after snapshots have been taken. Thus, if you snapshotted via `signal` at time `T1` then the next snapshot will be taken at `T1 + $interval`. But if you snapshot again at `T2` where `T2 > T1` then the next snapshot will be taken at `T2 + $interval`.
